src/gdf_table.cuh

```diff
@@ -1112,6 +1329,96 @@ private:
+                          cudaStream_t stream = 0) const
+  {
+    column_type * const i_data = static_cast<column_type*>(input_column->data);
+    column_type * const o_data = static_cast<column_type*>(output_column->data);
+    size_type num_rows = input_column->size;
+  
+    gdf_error gdf_status{GDF_SUCCESS};
+
+    // Gathering from one table to another
+    if (i_data != o_data) {
+        if (range_check) {
+            thrust::gather_if(thrust::cuda::par.on(stream),
+                    row_gather_map,
+                    row_gather_map + num_rows,
+                    row_gather_map,
+                    i_data,
+                    o_data,
+                    ValidRange<size_type>(0, num_rows));
+
+        } else {
+            thrust::gather(thrust::cuda::par.on(stream),
+                    row_gather_map,
+                    row_gather_map + num_rows,
+                    i_data,
+                    o_data);
+        }
+    } 
+    // Gather is in-place
+    else {
+        thrust::device_vector<column_type> remapped_copy(num_rows);
+        if (range_check) {
+            thrust::gather_if(thrust::cuda::par.on(stream),
+                           row_gather_map,
+                           row_gather_map + num_rows,
+                           row_gather_map,
+                           i_data,
+                           remapped_copy.begin(),
+                           ValidRange<size_type>(0, num_rows));
+        } else {
+            thrust::gather(thrust::cuda::par.on(stream),
+                           row_gather_map,
+                           row_gather_map + num_rows,
+                           i_data,
+                           remapped_copy.begin());
+        }
+        thrust::copy(thrust::cuda::par.on(stream),
+                remapped_copy.begin(),
+                remapped_copy.end(),
+                o_data);
+    }
+    //If gather is in-place
+    if ((input_column->valid == output_column->valid) &&
+            (input_column->valid != nullptr)) {
+        thrust::device_vector<gdf_valid_type> remapped_valid_copy(gdf_get_num_chars_bitmask(num_rows));
+        gather_valid<size_type>(
+                input_column->valid,
+                remapped_valid_copy.data().get(),
+                row_gather_map, 
+                num_rows, stream);
+        thrust::copy(thrust::cuda::par.on(stream),
+                remapped_valid_copy.begin(),
+                remapped_valid_copy.end(), output_column->valid);
+    }
+    //If both input and output columns have a non null valid pointer
+    else if (input_column->valid && output_column->valid) {
+        gather_valid<size_type>(
+                input_column->valid,
+                output_column->valid,
+                row_gather_map, 
+                num_rows, stream);
+    }
+    //If there is no input pointer then assume that all the valid masks are set to 1
+    else if ((!input_column->valid) && output_column->valid) {
```

Comment:
I'm not sure this case is possible. Even if it is, you could use `cudaMemset` instead of the overloaded `gather_valid` kernel.